### PR TITLE
6780 layers + analysis views

### PR DESCRIPTION
**Pull request by viddo**
Waypoint for #6780

Modeling of analysis on the editor frontend (i.e. integration with backend/cartodb.js TBD in following PR). Stopped here to avoid it to grow out of control.

To create a analysis model use (assumes to have a vis with at least one data layer) use the dev console for now:
```js
l = layerDefinitionsCollection.first();
m = analysisDefinitionsCollection.add({
  id: analysisDefinitionsCollection.nextId(l.get('source')),
  type: 'trade-area',
  params: {
    source: l.get('source'),
    kind: 'walk',
    time: 300
  }
});
l.set('source', m.id);

// point to another's layers source
l = layerDefinitionsCollection.first();
l1 = layerDefinitionsCollection.at(1);
m = analysisDefinitionsCollection.add({
  id: analysisDefinitionsCollection.nextId(l.get('source')),
  type: 'trade-area',
  params: {
    source: l1.get('source'),
    kind: 'walk',
    time: 300
  }
});
l.set('source', m.id);
```

Detailed scope:
- [x] base views for layer definitions
- [x] models + collections for analysis
- [x] create analysis model representation for layers w/ table or query
- [x] maintain table_name+query props (for backward compability with existing layer)
- [x] set a letter representation for each layer
- [x] `analysisDefinitionsCollection.getNextId(layer)` to get next id for a new analysis node of a given layer
- [x] basemap should not have letter representation
- [x] render source analysis node for own layer (i.e. no other refs)
- [x] render analysis node as reference if used by another layer
- [x] create view for an analysis node that uses a previous node as input (e.g. trade-area, implement model using definition proposal for data)

cc @xavijam @javierarce

### CODE REVIEW

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js [hunk 1]
```diff
@@ -1,18 +1,53 @@
...
+      d.options
+    );
+    var attrs = _.defaults(
+      {
+        options: o
+      },
+      _.omit(d, ['options'])
+    );
+
+    if (attrs.kind !== 'tiled') {
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: Why `kind` and not `type`? In CartoDB.js we normally use `type`, but I've also seen that sometimes we fallback to `kind` just in case... Probably something that was changed in the past. Can we stick to `type`?

cc: @juanignaciosl

File: lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    this._configModel = options.configModel;
+  },
+
+  /**
+   * Get next id in sequence of given source id.
+   * @param {String} sourceId e.g. 'B2'
+   * @return {String} e.g. 'B3'
+   */
+  nextId: function (sourceId) {
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js]: Is this being used?

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js (hunk 1 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: It's the name of the attribute returned by the layer data from backend. 

Also, to make it more confusing, there exists both kind and type, and they have different values ("tiled" vs "Tiled", "carto" vs. "CartoDB")…  not sure if it's possible to change that since there's code relying on both these attributes.

File: lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js (hunk 2 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js]: Yes, didn't you read the PR description (see the snippet example)? :wink:

File: lib/assets/test/spec/cartodb3/data/analysis-definitions/analysis-source-definition-model.spec.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      table_name: 'foo_bar'
+    });
+  });
+
+  describe('.toJSON', function () {
+    beforeEach(function () {
+      this.json = this.model.toJSON();
+    });
+
+    it('should generate a json with params', function () {
```
Comment (alonsogarciapablo) [inline lib/assets/test/spec/cartodb3/data/analysis-definitions/analysis-source-definition-model.spec.js]: Why not assert that:

```
expect(this.json).toEqual({
  id: 'a0',
  type: 'source',
  params: {
    query: 'SELECT * FROM foo_bar'
  }
});
```

? (It's a bit more readable IMO)

Comment (viddo) [inline lib/assets/test/spec/cartodb3/data/analysis-definitions/analysis-source-definition-model.spec.js]: Not sure toEqual can do deep-equals, but good point, I'll give it a try

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js (hunk 1 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: Btw, I believe `kind` is the type the backend uses internally, `type` is the custom prop that exists in `attrs.options` prop, which is set/get by frontend/cartodb.js.

File: lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js (hunk 2 above)
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js]: I actually didn't read well enough, sorry. :+1:

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js (hunk 1 above)
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: Let's see what @juanignaciosl has to say...

Regarding this issue, I opened CartoDB/cartodb#4348 a while back.

File: lib/assets/test/spec/cartodb3/data/analysis-definitions/analysis-source-definition-model.spec.js (hunk 3 above)
Comment (viddo) [inline lib/assets/test/spec/cartodb3/data/analysis-definitions/analysis-source-definition-model.spec.js]: easier to read (and verified working), downside would be harder-to-read failed-test-output, e.g.:
`Expected Object({ id: 'a0', type: 'trade-area', params: Object({ kind: 'walk', query: 'SELECT * FROM foo_bar', time: 300 }) }) to equal Object({ id: 'a0', type: 'trade-area', params: Object({ query: 'SELECT * FROM foo_bar', kind: 'car', time: 300 }) }).`

Comment (alonsogarciapablo) [inline lib/assets/test/spec/cartodb3/data/analysis-definitions/analysis-source-definition-model.spec.js]: Yes, I use that kind of assertions in CartoDB.js. Agree that ouput is hard to read when test fails. Your call then.

Comment (viddo) [inline lib/assets/test/spec/cartodb3/data/analysis-definitions/analysis-source-definition-model.spec.js]: Thx to your proposal it became obvious that the serialized object was actually wrong(!) fixed in  29d9296. Also, the object should be a (fairly) shallow anyway, so going with your proposal. :+1:

Comment (alonsogarciapablo) [inline lib/assets/test/spec/cartodb3/data/analysis-definitions/analysis-source-definition-model.spec.js]: :clap:

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js (hunk 1 above)
Comment (juanignaciosl) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: It's `kind` internally (`layers.kind` column). It's messy:

```
./app/controllers/carto/api/layer_presenter.rb:          with_kind_as_type(base_poro(@layer)).symbolize_keys
./app/controllers/carto/api/layer_presenter.rb:      def with_kind_as_type(attributes)
./app/controllers/carto/api/layer_presenter.rb:        decorate_with_data(attributes.merge(type: attributes.delete('kind')), @decoration_data)
./app/controllers/carto/api/layer_vizjson_adapter.rb:          # TODO: kind should be renamed to type
./app/controllers/carto/api/layer_vizjson_adapter.rb:          # rename once a new layer presenter is written. See CartoDB::LayerModule::Presenter#with_kind_as_type
./app/models/carto/visualization.rb:    !(kind_raster? || type_slide?)
./app/models/layer/presenter.rb:          with_kind_as_type(layer.public_values).symbolize_keys
./app/models/layer/presenter.rb:      def with_kind_as_type(attributes)
./app/models/layer/presenter.rb:        decorate_with_data(attributes.merge(type: attributes.delete('kind')), @decoration_data)
./app/services/carto/visualizations_export_service.rb:      data['kind'] = layer_kind_from_type(exported_layer['type'])
./app/services/carto/visualizations_export_service.rb:    def layer_kind_from_type(exported_layer_type)
./app/services/carto/visualizations_export_service.rb:        kind = layer_kind_from_type(layer["type"])
```

IMHO `type` makes more sense. I don't know which one came first, but currently `kind` looks like a backend implementation detail leaked everywhere. I'd vote for using `type` everywhere for the new code.

My only concern is that `layers.options` contains _some_ (not too many) occurrences of `kind`. I can't know if they were generated at the frontend or they're caused by the "round trip" to the database.

I'd try to make all new code based on `type`, with lower cased values, and `carto` instead `cartodb` (`cartodb` is not even a current `kind` value).

File: lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+var SOURCE_ID_REGEX = /^([a-zA-Z]+)(\d+)$/;
+
+/**
+ * Collection of analysis definitions
+ */
+module.exports = Backbone.Collection.extend({
+
+  model: function (d, opts) {
+    var self = opts.collection;
+    // var Klass = TYPE_TO_ANALYSIS_DEFINITION_MODEL_MAP[d.type] || cdb.core.Model;
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js]: Do we need this comment?

File: lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+};
+var SOURCE_ID_REGEX = /^([a-zA-Z]+)(\d+)$/;
+
+/**
+ * Collection of analysis definitions
+ */
+module.exports = Backbone.Collection.extend({
+
+  model: function (d, opts) {
+    var self = opts.collection;
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js]: For me this is confusing, when `var self = opts.collection;`?

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js (hunk 1 above)
Comment (juanignaciosl) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: Oh, I hadn't read @alonsogarciapablo proposal on CartoDB/cartodb#4348, sorry. If you prefer kind I don't really mind, I thought `type` was ubiquitous. And I 100% agree on not being an `option`.

File: lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js (hunk 5 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js]: because `this` can't be used (depending on how the collection is called it might be the collection, window, or something else), thus using the `self` (in the same way that `var self = this` is used on some occasions with closures) as a way to always have the context for the own collection. 

This is used in the other collections as well, so if we should update them too. How would you change this to make it clearer?

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js (hunk 1 above)
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: My bad. I actually prefer `type` cause that's CartoDB.js is using. That is issue is a bit old (I'll update it). Not using `kind` in the frontend / cartodb.js is a good start IMO.

File: lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js (hunk 5 above)
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js]: Yeah, I got it, but not sure... `analysisCollection`?
Anyway, not really important -> _optional_

**Comment by xavijam** [comment]
Looks promising guys, great job :clap:

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js (hunk 1 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: Sounds reasonable, I'll change this to use `type` then, thx for the conversation/explanations :+1:

File: lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js (hunk 4 above)
Comment (javierarce) [inline lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js]: Nope.

**Comment by juanignaciosl** [comment]
Comment thread has been lost. I think we'll use `type`, lowercased values and `carto` instead of `cartodb`, isn't it? Please open issues if you ever get something wrong from the backend.

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js (hunk 1 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: From @juanignaciosl 
> I think we'll use type, lowercased values and carto instead of cartodb, isn't it? Please open issues if you ever get something wrong from the backend.

Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: But as @alonsogarciapablo points out in the referenced issue:

> `Layer.all().map{|layer| layer.options['type']}.uniq // => ["Tiled", "CartoDB", nil, "torque", "Plain"]`

The values are _not_ lowercase, wouldn't changing that break backward compability? :neutral_face:

Comment (juanignaciosl) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: I thought we were speaking about making some changes, @alonsogarciapablo even talked about not having `type` inside options.

TL;DR: _if we are going to do any changes, I'd_ use type, lowercased values and carto instead of cartodb. 

Some data (taken from ded02):

```
cartodb_staging=# select kind, count(1) from layers group by kind;
    kind    | count
------------+-------
 tiled      |  1854
 gmapsbase  |     5
 torque     |    23
 carto      |  1647
 background |     6
(5 rows)

cartodb_staging=# select options::json->>'type' options_type, count(1) from layers group by options_type;
 options_type | count
--------------+-------
              |  1578
 Tiled        |  1467
 GMapsBase    |     1
 CartoDB      |   460
 Plain        |     6
 torque       |    23
(6 rows)
```

Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: Fair enough, but I think this kind of change is not in the scope of this PR, right? 
For now I changed it to use `options.type` (as previously suggested) and to not have to do this kind of change that could possibly break other stuff. 
I'd suggest we stop here, and if we want to change the location of type out of options to do that separately. Sounds ok?

Comment (juanignaciosl) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: :ok_hand:

Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: Agree with latest comment from @viddo. We don't need to change this now.
